modules/dlm/src/main/java/org/elasticsearch/dlm/ModifyDataLifecycleService.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.dlm;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateAckListener;
+import org.elasticsearch.cluster.ClusterStateTaskExecutor;
+import org.elasticsearch.cluster.ClusterStateTaskListener;
+import org.elasticsearch.cluster.SimpleBatchedAckListenerTaskExecutor;
+import org.elasticsearch.cluster.metadata.DataLifecycle;
+import org.elasticsearch.cluster.metadata.DataStream;
+import org.elasticsearch.cluster.metadata.IndexAbstraction;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.cluster.service.MasterServiceTaskQueue;
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.core.Tuple;
+
+import java.util.List;
+
+/**
+ * This service modifies the data lifecycle of an existing data stream. It creates and batches the cluster state update tasks
+ * that will change the data stream metadata in the cluster state.
+ */
+public class ModifyDataLifecycleService {
+
+    private final MasterServiceTaskQueue<ModifyLifecycleTask> taskQueue;
+
+    @Inject
+    public ModifyDataLifecycleService(ClusterService clusterService) {
+        ClusterStateTaskExecutor<ModifyLifecycleTask> executor = new SimpleBatchedAckListenerTaskExecutor<>() {
+
+            @Override
+            public Tuple<ClusterState, ClusterStateAckListener> executeTask(
+                ModifyLifecycleTask modifyLifecycleTask,
+                ClusterState clusterState
+            ) {
+                return new Tuple<>(
+                    modifyLifecycle(clusterState, modifyLifecycleTask.dataStreamNames(), modifyLifecycleTask.dataLifecycle()),
+                    modifyLifecycleTask
+                );
+            }
+        };
+        this.taskQueue = clusterService.createTaskQueue("modify-lifecycle", Priority.LOW, executor);
```

Comment:
Shall we go with at least `NORMAL` (I'm even thinking `HIGH` given this could allow faster deletion of indices or skipping deleting them)?

Reply:
Since deleting a data stream has a `HIGH` priority I think we should indeed use priority `HIGH`.